### PR TITLE
NUTCH-3161 Address Sonarcloud High and Medium Security Hotspots

### PR DESCRIPTION
**Pull request by lewismc**
PR to address [NUTCH-3161](https://issues.apache.org/jira/browse/NUTCH-3161). This patch addresses following Security Hotspota
 
### High
(false positives): Exclude plugin resource directories from analysis in `sonar-project.properties`. No Java code lives in `conf`, `data`, or `sample` under `src/plugin/**`, so these paths are excluded from scanning., and 

### Medium
* ParseOutputFormat (`src/java/.../ParseOutputFormat.java`): Replace regex `" *, *"` for `db.parsemeta.to.crawldb` with comma-split + trim in `getParseMetaToCrawlDBKeys()`. Add `TestParseOutputFormat` with tests for empty, single, comma-separated, trim, and empty-segment handling.
* HtmlParser (parse-html plugin): Remove `metaPattern`, `charsetPattern`, and `charsetPatternHTML5`. Use linear string parsing in `extractCharsetFromMeta()` / `extractCharsetValue()` for HTML4 and HTML5 meta charset detection. Add `testExtractCharsetFromMeta` in existing `TestHtmlParser`.
* JSParseFilter (parse-js plugin): Remove `STRING_PATTERN` and `URI_PATTERN`. Use `extractQuotedStrings()` and `looksLikeUri()` (linear scan / simple checks). Add `testExtractQuotedStrings` and `testLooksLikeUri` in existing `TestJSParseFilter`.
* UrlValidator (urlfilter-validator plugin): Remove `URL_PATTERN` and `AUTHORITY_PATTERN`. Use `java.net.URI` for URL structure and `parseAuthority()` for host/port. Remove unused `isBlankOrNull`. Add `testParseAuthority` in existing`TestUrlValidator`.

Thanks for any review.

### CODE REVIEW

**Comment by lewismc** [comment]
Confirmed this PR addresses the [security hotspots](https://sonarcloud.io/project/security_hotspots?id=apache_nutch&pullRequest=904) and passes tests.

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 1]
```diff
@@ -93,6 +86,82 @@ public class HtmlParser implements Parser {
...
+    return end > start ? s.substring(start, end) : null;
+  }
+
+  /**
+   * Finds charset from HTML string using linear scans only (no backtracking regex).
+   * Checks meta http-equiv Content-Type then HTML5 meta charset.
+   * Package-private for unit testing.
+   */
+  static String extractCharsetFromMeta(String str) {
+    String lower = str.toLowerCase();
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:133]: Should add `Locale.ROOT`.

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 2]
```diff
@@ -102,20 +171,7 @@ private static String sniffCharacterEncoding(byte[] content) {
     // {U+0041, U+0082, U+00B7}.
     String str = new String(content, 0, length, StandardCharsets.US_ASCII);
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:172]: The content bytes are converted into a String assuming ASCII encoding. If a "hand-crafted" scanner is used, could directly operate on the bytes avoiding the conversion to a string. Alternatively, could use a CharSequence wrapping the bytes, cf. [ByteArrayCharSequence](https://github.com/commoncrawl/nutch/blob/cc/src/java/org/commoncrawl/util/ByteArrayCharSequence.java). But this can be an improvement for later.

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 3]
```diff
@@ -93,6 +86,82 @@ public class HtmlParser implements Parser {
...
+    while (true) {
+      int metaStart = lower.indexOf(META_TAG_START, pos);
+      if (metaStart < 0) {
+        break;
+      }
+      int tagEnd = str.indexOf('>', metaStart);
+      if (tagEnd < 0) {
+        break;
+      }
+      String tagContent = str.substring(metaStart, tagEnd);
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:144]: Assuming that indexes in the lower-case string and the original one are the same might be dangerous, for example, in German there is the pair "ß" <> "SS" (would held for uppercasing the string). See [Unicode TR #21 Case Mappings](https://www.unicode.org/L2/L1999/99190.htm): "Case mappings may produce strings of different length than the original."

Comment (sebastian-nagel) [reply src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:144]: Of course, if we strictly ensure that the method is used on String only including ASCII characters and using always the root locale, this should be safe.

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 4]
```diff
@@ -93,6 +86,82 @@ public class HtmlParser implements Parser {
...
+      String tagLower = tagContent.toLowerCase();
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:145]: Same: must use `Locale.ROOT`.

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 5]
```diff
@@ -64,15 +62,10 @@ public class HtmlParser implements Parser {
...
-  private static Pattern metaPattern = Pattern.compile(
-      "<meta\\s+([^>]*http-equiv=(\"|')?content-type(\"|')?[^>]*)>",
-      Pattern.CASE_INSENSITIVE);
-  private static Pattern charsetPattern = Pattern.compile(
-      "charset=\\s*([a-z][_\\-0-9a-z]*)", Pattern.CASE_INSENSITIVE);
-  private static Pattern charsetPatternHTML5 = Pattern.compile(
-      "<meta\\s+charset\\s*=\\s*[\"']?([a-z][_\\-0-9a-z]*)[^>]*>",
-      Pattern.CASE_INSENSITIVE);
+  private static final String META_TAG_START = "<meta";
+  private static final String CHARSET_EQ = "charset=";
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:66]: The regex allows white space between "charset" and "=".

File: src/java/org/apache/nutch/parse/ParseOutputFormat.java [hunk 6]
```diff
@@ -177,8 +194,8 @@ public RecordWriter<Text, Parse> getRecordWriter(TaskAttemptContext context)
     Path data = new Path(new Path(out, ParseData.DIR_NAME), name);
     Path crawl = new Path(new Path(out, CrawlDatum.PARSE_DIR_NAME), name);
 
-    final String[] parseMDtoCrawlDB = conf.get("db.parsemeta.to.crawldb", "")
```
Comment (sebastian-nagel) [inline src/java/org/apache/nutch/parse/ParseOutputFormat.java:180]: It's a string from the configuration, it's controlled by the user and should be short. It's parsed once, so this is for sure not critical.

Apart from that should simply rely on [getTrimmedStrings](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/conf/Configuration.html#getTrimmedStrings(java.lang.String,java.lang.String...)).

File: src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java [hunk 7]
```diff
@@ -93,6 +86,82 @@ public class HtmlParser implements Parser {
...
+   * Extracts charset value from a string like "charset=utf-8" or "charset = utf-8".
+   * Uses linear scan to avoid ReDoS. Value must start with [a-z] and contain only [a-z0-9_-].
+   */
+  private static String extractCharsetValue(String s, int fromIndex) {
+    int idx = s.indexOf(CHARSET_EQ, fromIndex);
+    if (idx < 0) {
+      return null;
+    }
+    int start = idx + CHARSET_EQ.length();
+    while (start < s.length() && (s.charAt(start) == ' ' || s.charAt(start) == '\t')) {
```
Comment (sebastian-nagel) [inline src/plugin/parse-html/src/java/org/apache/nutch/parse/html/HtmlParser.java:99]: `\\s` matches more characters than blank (U+0020) and the tab character.

**Review by sebastian-nagel — CHANGES_REQUESTED** [review]
Hi @lewismc, thanks!

So far, this is only a partial review. See the inline comments. Without having tried it, I expect some regressions, e.g. when extracting the charset from `<meta charset    = "UTF-8" />`. Replacing regular expressions by hand-written scanner code is difficult and makes code maintenance much more difficult. Maybe use a scanner generator instead, e.g. Ragel? Alternatively, we could use [RE2/J](https://github.com/google/re2j) or [dk.brics.automaton](https://www.brics.dk/automaton/) (see urlfilter-automaton) which are faster (and more safe) because they do not backtrack.

In addition, the URL validator needs a more substantial fix, see NUTCH-2986.

**Comment by lewismc** [comment]
Thank you for the sincere review @sebastian-nagel I appreciate it. This PR was too adventurous and too large. I'll close it and move to these issues later on.
